common/setups/rasr/gmm_system.py

```diff
@@ -1007,6 +1105,37 @@ def run_vtln_sat_step(self, step_args):
                     **step_args.sdm_args,
                 )
 
+    def run_recognition_step(self, step_args):
```

Comment:
Ive been writing "see above" now so many times that I would suggest to remove code duplication by moving out the two loops and calls to `self.recognition` into a separate function. Or maybe this `run_recognition_step` is already sufficient and you just call this in each training step.